File: chapter2/javascript/linkedlist.js
class Node {
    constructor(val) {
        this.val = val;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    printList() {
        let node = this.head;
        let res = "";
        while (node) {
            res += node.val + " => ";
            node = node.next;
        }
        console.log(res);
    }

    toArray() {
        let res = [];
        let node = this.head;
        while (node) {
            res.push(node.val);
            node = node.next;
        }
        return res;
    }
    append(val) {
        let node = new Node(val);
        // if it's an empty list
        if (!this.head) {
            this.head = node;
            this.tail = node;
        } else {
            this.tail.next = node;
            this.tail = node;
        }
        this.size++;
    }

    prepend(val) {
        let node = new Node(val);
        node.next = this.head;
        this.head = node;
        this.size++;
    }

    remove(index) {
        if (index == 0) {
            this.head = this.head.next;
        } else {
            let node = this.head;
            // loop till you reach the previous node to the target, the index here is more like "remaining steps"
            while (index - 1) {
                node = node.next;
                index--;
            }
            if (!node.next.next) {
                this.tail = node;
            }
            node.next = node.next.next;

        }
        this.size--;
    }

    pop() {
        this.remove(this.size - 1);
    }
    popFirst() {
        this.remove(0);
    }
}

module.exports = LinkedList;

// TESTS 

// let my_list = new LinkedList();


// my_list.append(5);
// my_list.printList();
// my_list.append(8);
// my_list.printList();
// my_list.append(7);
// my_list.printList();
// my_list.remove(1);
// my_list.printList();
// my_list.prepend(10);
// my_list.printList();
// my_list.append(22);
// my_list.printList();
// my_list.popFirst();
// my_list.pop();

// my_arr = my_list.toArray();

// my_list.printList();

// console.log(my_arr);

File: chapter2/python/linkedlist.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def reverse_list(self):
        prev = None
        curr = self.head
        while(curr != None):
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        self.head = prev

    def printList(self):
        node = self.head
        res = ""
        while(node):
            res += str(node.val) + " => "
            node = node.next
        print(res)

    def toArray(self):
        res = []
        node = self.head
        while node:
            res.append(node.val)
            node = node.next
        return res

    def append(self, val):
        node = Node(val)
        # if it's an empty list
        if not self.head or not self.tail:
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            self.tail = node
        self.size += 1

    def prepend(self, val):
        node = Node(val)
        # if it's an empty list
        if not self.head or not self.tail:
            self.head = node
            self.tail = node
        else:
            node.next = self.head
            self.head = node
        self.size += 1

    def appendNode(self, node):
        if not self.head or not self.tail:
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            self.tail = node
        self.size += 1

    def remove(self, index):
        if index == 0:
            self.head = self.head.next
        else:
            node = self.head
            # loop till you reach the previous node to the target, the index here is more like "remaining steps"
            while (index-1):
                node = node.next
                index -= 1
            if not node.next.next:
                self.tail = node
            node.next = node.next.next
        self.size -= 1

    def pop(self):
        return self.remove(self.size-1)

    def popFirst(self):
        return self.remove(0)

# TESTS

# my_list = LinkedList()


# my_list.append(5)
# my_list.printList()
# my_list.append(8)
# my_list.printList()
# my_list.append(7)
# my_list.printList()
# my_list.remove(1)
# my_list.printList()
# my_list.prepend(10)
# my_list.printList()
# my_list.append(22)
# my_list.printList()
# my_list.popFirst()
# my_list.pop()

# my_arr = my_list.toArray()

# my_list.printList()

# print(my_arr)
